content/intro-to-storybook/ember/en/composite-component.md

```diff
@@ -0,0 +1,301 @@
+
+![empty and loading tasks](/intro-to-storybook/tasklist-states-2.png)
+
+## Get setup
+
+A composite component isn’t much different than the basic components it contains. Create a `TaskList` template, component and an accompanying story file: `app/components/TaskList.hbs`, `app/components/TaskList.js` and `app/components/TaskList.stories.js`.
+
+Start with a rough implementation of the `TaskList`. You’ll need to import the `Task` component from earlier and pass in the attributes and actions as inputs.
+
+```handlebars
+
+{{!-- app/components/TaskList.hbs--}}
+{{#if (eq @loading true)}}
+  <div class="list-items">loading</div>
+{{/if}}
+
+{{#if this.emptyTasks}}
+  <div class="list-items">empty</div>
+{{/if}}
+
+{{#each @tasks as |task|}}
+ {{Task task=task pinTask=(action "taskListOnPinTask") archiveTask=(action "taskListOnArchiveTask")}}
+{{/each}}
+```
+
+Then we'll add the following to `TaskList.js`:
+
+```javascript
+// app/components/TaskList.js
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+export default class TaskList extends Component {
+  constructor() {
+    super(...arguments);
+  }
+  get emptyTasks() {
+    return this.args.tasks.length === 0 && !this.args.loading;
+  }
+  // actions to emulate the click handlers to call the actions passed down to the component
+  @action
+  taskListOnPinTask(taskID) {
+    if (this.args.pinTask) {
+      this.args.pinTask(taskID);
+    }
+  }
+  @action
+  taskListOnArchiveTask(taskID) {
+    if (this.args.archiveTask) {
+      this.args.archiveTask(taskID);
+    }
+  }
+}
+```
+
+Next create `Tasklist`’s test states in the story file.
+
+```javascript
+// app/components/TaskList.stories.js
+import { hbs } from 'ember-cli-htmlbars';
+import { taskData, actionsData } from './Task.stories';
+
+export default {
+  title: 'TaskList',
+  component: 'TaskList',
+  excludeStories: /.*Data$/,
+};
+
+export const defaultTasksData = [
+  { ...taskData, id: '1', title: 'Task 1' },
+  { ...taskData, id: '2', title: 'Task 2' },
+  { ...taskData, id: '3', title: 'Task 3' },
+  { ...taskData, id: '4', title: 'Task 4' },
+  { ...taskData, id: '5', title: 'Task 5' },
+  { ...taskData, id: '6', title: 'Task 6' },
+];
+
+export const withPinnedTasksData = [
+  ...defaultTasksData.slice(0, 5),
+  { id: '6', title: 'Task 6 (pinned)', state: 'TASK_PINNED' },
+];
+
+export const Default = () => ({
+  template: hbs`<div style="padding: 3rem">{{TaskList tasks=tasks pinTask=(action taskActions.onPinTask) archiveTask=(action taskActions.onArchiveTask)}}</div>`,
+  context: {
+    tasks: defaultTasksData,
+    taskActions: {
+      ...actionsData,
+    },
+  },
+});
+
+export const WithPinnedTasks = () => ({
+  template: hbs`<div style="padding: 3rem">{{TaskList tasks=tasks pinTask=(action taskActions.onPinTask) archiveTask=(action taskActions.onArchiveTask)}}</div>`,
+  context: {
+    tasks: withPinnedTasksData,
+    taskActions: {
+      ...actionsData,
+    },
+  },
```

Comment:
The context should mimic the backing class in ember. Actions are no longer contained in an `actions` hash but rather just methods in the class. The final `context` should look similar:

```
context: {
  tasks: [...],
  pinTask: () => {},
  archiveTask: () => {}
}
```

but you are still able to compose it together:

```
// either
context: {
  tasks: whatEverTasksYouNEed,
  ...actionsHash
};

// --- or ---

context: {
  ...tasks,
  ...actionsHash
}
```